File: golang/binary-tree-general/construct_binary_tree_postorder.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	inorderMap := make(map[int]int)

	for i, val := range inorder {
		inorderMap[val] = i
	}

	var helper func(inLeft, inRight int) *TreeNode

	helper = func(inLeft, inRight int) *TreeNode {
		if inLeft > inRight {
			return nil
		}

		rootVal := postorder[len(postorder)-1]
		postorder = postorder[:len(postorder)-1]

		root := &TreeNode{Val: rootVal}

		index := inorderMap[rootVal]

		root.Right = helper(index+1, inRight)
		root.Left = helper(inLeft, index-1)

		return root
	}

	return helper(0, len(inorder)-1)
}

File: python/binary-tree-general/construct_binary_tree_postorder.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val: int = 0, left: None= None, right: None = None) -> None:
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) ->  Optional[TreeNode]:
        inorder_map = {
            val: idx for idx, val in enumerate(inorder)
        }


        def helper(in_left: int, in_right: int) -> Optional[TreeNode]:
            if in_left > in_right:
                return None


            root_val = postorder.pop()
            root = TreeNode(root_val)

            index = inorder_map[root_val]

            root.right = helper(index + 1, in_right)
            root.left = helper(in_left, index - 1)

            return root

        return helper(0, len(inorder) - 1)
